src/main/java/com/example/petbuddybackend/middleware/interceptor/ValidChatRoomAccessInterceptor.java

```diff
@@ -1,59 +1,128 @@
 import org.springframework.stereotype.Component;
 
+@Slf4j
 @Component
 @RequiredArgsConstructor
-public class ValidChatRoomAccessInterceptor implements ChannelInterceptor {
+public class ValidChatRoomAccessInterceptor implements ChannelInterceptor, ApplicationContextAware {
+
+    private static final String CHAT = "Chat";
+    private static final String USER_NOT_IN_CHAT_MESSAGE = "User %s is not in chat %d";
 
     @Value("${url.chat.topic.subscribe-prefix}")
     private String URL_CHAT_TOPIC_BASE;
 
+    @Value("${url.exception.topic.send-url}")
+    public String EXCEPTIONS_PATH = "/topic/exceptions";
+
     @Value("${url.chat.topic.chat-id-pos}")
     private int CHAT_ID_INDEX_IN_TOPIC_URL;
 
     @Value("${header-name.role}")
     private String ROLE_HEADER_NAME;
 
     private final ChatRoomRepository chatRoomRepository;
+    private final BlockService blockService;
+    private SimpMessagingTemplate simpMessagingTemplate;
+    private ApplicationContext applicationContext;
 
+    /**
+     * Required for lazy loading {@link SimpMessagingTemplate} as it is initialized after this interceptor
+     */
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+        this.applicationContext = applicationContext;
+    }
 
     @Override
     public Message<?> preSend(Message<?> message, MessageChannel channel) {
+        // Lazy load SimpMessagingTemplate
+        if (simpMessagingTemplate == null) {
+            simpMessagingTemplate = applicationContext.getBean(SimpMessagingTemplate.class);
+        }
+
         StompHeaderAccessor accessor = StompHeaderAccessor.wrap(message);
         String destination = accessor.getDestination();
 
-        if(HeaderUtils.destinationStartsWith(URL_CHAT_TOPIC_BASE, destination)) {
-            handleMessageTopic(accessor, destination);
+        if(HeaderUtils.destinationStartsWith(URL_CHAT_TOPIC_BASE, destination) && !accessPermitted(accessor, destination)) {
+            return null;
         }
 
         return message;
     }
 
-    private void handleMessageTopic(StompHeaderAccessor accessor, String destination) {
+    private boolean accessPermitted(StompHeaderAccessor accessor, String destination) {
         StompCommand command = accessor.getCommand();
         Role role = Role.valueOf(accessor.getFirstNativeHeader(ROLE_HEADER_NAME));
 
         if(!StompCommand.SUBSCRIBE.equals(command) && !StompCommand.SEND.equals(command)) {
-            return;
+            return true;
         }
 
         String username = HeaderUtils.getUser(accessor);
         Long chatId = extractChatId(destination);
+        String sessionId = accessor.getSessionId();
+
+        return validateChatPermission(chatId, username, sessionId, role);
+    }
 
+    private boolean validateChatPermission(Long chatId, String username, String sessionId, Role role) {
+        try {
+            ChatRoom chatRoom = chatRoomRepository.findById(chatId).orElseThrow(
```

Comment:
mozesz sprobowac wstrzyknac sobie chatService i wtedy moznaby bylo część z tego tam przenieść. Ja to wywalilem przez circular dependencies, ale potem zmienilem troche kod, wiec moze juz zadziala

Reply:
Oo faktycznie działa. Podmieniłem